File: src/hooks/session-end.ts
/**
 * ContextStream session-end Hook - Finalize session on stop
 *
 * Stop hook that finalizes the session, saving the full transcript
 * and marking the session as complete in ContextStream.
 *
 * Usage:
 *   npx @contextstream/mcp-server hook session-end
 *
 * Input (stdin): JSON with session_id, transcript_path, cwd
 * Output (stdout): None (silent finalization)
 * Exit: Always 0
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { homedir } from "node:os";

const ENABLED = process.env.CONTEXTSTREAM_SESSION_END_ENABLED !== "false";
const SAVE_TRANSCRIPT = process.env.CONTEXTSTREAM_SESSION_END_SAVE_TRANSCRIPT !== "false";

let API_URL = process.env.CONTEXTSTREAM_API_URL || "https://api.contextstream.io";
let API_KEY = process.env.CONTEXTSTREAM_API_KEY || "";
let WORKSPACE_ID: string | null = null;
let PROJECT_ID: string | null = null;

interface HookInput {
  session_id?: string;
  transcript_path?: string;
  cwd?: string;
  hook_event_name?: string;
  reason?: string; // Why session ended (user_exit, timeout, etc.)
}

interface McpConfig {
  mcpServers?: {
    contextstream?: {
      env?: {
        CONTEXTSTREAM_API_KEY?: string;
        CONTEXTSTREAM_API_URL?: string;
      };
    };
  };
}

interface LocalConfig {
  workspace_id?: string;
  project_id?: string;
}

interface TranscriptMessage {
  role: string;
  content: string;
  timestamp: string;
  tool_calls?: unknown;
  tool_results?: unknown;
}

interface TranscriptStats {
  messageCount: number;
  toolCallCount: number;
  duration: number; // seconds
  filesModified: string[];
  messages: TranscriptMessage[];
  startedAt: string;
}

function loadConfigFromMcpJson(cwd: string): void {
  let searchDir = path.resolve(cwd);

  for (let i = 0; i < 5; i++) {
    if (!API_KEY) {
      const mcpPath = path.join(searchDir, ".mcp.json");
      if (fs.existsSync(mcpPath)) {
        try {
          const content = fs.readFileSync(mcpPath, "utf-8");
          const config = JSON.parse(content) as McpConfig;
          const csEnv = config.mcpServers?.contextstream?.env;
          if (csEnv?.CONTEXTSTREAM_API_KEY) {
            API_KEY = csEnv.CONTEXTSTREAM_API_KEY;
          }
          if (csEnv?.CONTEXTSTREAM_API_URL) {
            API_URL = csEnv.CONTEXTSTREAM_API_URL;
          }
        } catch {
          // Continue
        }
      }
    }

    if (!WORKSPACE_ID || !PROJECT_ID) {
      const csConfigPath = path.join(searchDir, ".contextstream", "config.json");
      if (fs.existsSync(csConfigPath)) {
        try {
          const content = fs.readFileSync(csConfigPath, "utf-8");
          const csConfig = JSON.parse(content) as LocalConfig;
          if (csConfig.workspace_id && !WORKSPACE_ID) {
            WORKSPACE_ID = csConfig.workspace_id;
          }
          if (csConfig.project_id && !PROJECT_ID) {
            PROJECT_ID = csConfig.project_id;
          }
        } catch {
          // Continue
        }
      }
    }

    const parentDir = path.dirname(searchDir);
    if (parentDir === searchDir) break;
    searchDir = parentDir;
  }

  // Check home .mcp.json
  if (!API_KEY) {
    const homeMcpPath = path.join(homedir(), ".mcp.json");
    if (fs.existsSync(homeMcpPath)) {
      try {
        const content = fs.readFileSync(homeMcpPath, "utf-8");
        const config = JSON.parse(content) as McpConfig;
        const csEnv = config.mcpServers?.contextstream?.env;
        if (csEnv?.CONTEXTSTREAM_API_KEY) {
          API_KEY = csEnv.CONTEXTSTREAM_API_KEY;
        }
        if (csEnv?.CONTEXTSTREAM_API_URL) {
          API_URL = csEnv.CONTEXTSTREAM_API_URL;
        }
      } catch {
        // Ignore
      }
    }
  }
}

function parseTranscriptStats(transcriptPath: string): TranscriptStats {
  const stats: TranscriptStats = {
    messageCount: 0,
    toolCallCount: 0,
    duration: 0,
    filesModified: [],
    messages: [],
    startedAt: new Date().toISOString(),
  };

  if (!transcriptPath || !fs.existsSync(transcriptPath)) {
    return stats;
  }

  try {
    const content = fs.readFileSync(transcriptPath, "utf-8");
    const lines = content.split("\n");
    let firstTimestamp: Date | null = null;
    let lastTimestamp: Date | null = null;
    const modifiedFiles = new Set<string>();

    for (const line of lines) {
      if (!line.trim()) continue;

      try {
        const entry = JSON.parse(line) as {
          type?: string;
          role?: string;
          name?: string;
          content?: string;
          input?: { file_path?: string; notebook_path?: string };
          timestamp?: string;
        };

        const msgType = entry.type || "";
        const timestamp = entry.timestamp || new Date().toISOString();

        // Track timestamps
        if (entry.timestamp) {
          const ts = new Date(entry.timestamp);
          if (!firstTimestamp || ts < firstTimestamp) {
            firstTimestamp = ts;
            stats.startedAt = entry.timestamp;
          }
          if (!lastTimestamp || ts > lastTimestamp) {
            lastTimestamp = ts;
          }
        }

        // Count and capture messages
        if (msgType === "user" || entry.role === "user") {
          stats.messageCount++;
          const userContent = typeof entry.content === "string" ? entry.content : "";
          if (userContent) {
            stats.messages.push({
              role: "user",
              content: userContent,
              timestamp,
            });
          }
        } else if (msgType === "assistant" || entry.role === "assistant") {
          stats.messageCount++;
          const assistantContent = typeof entry.content === "string" ? entry.content : "";
          if (assistantContent) {
            stats.messages.push({
              role: "assistant",
              content: assistantContent,
              timestamp,
            });
          }
        } else if (msgType === "tool_use") {
          stats.toolCallCount++;
          const toolName = entry.name || "";
          const toolInput = entry.input || {};

          // Track file modifications
          if (["Write", "Edit", "NotebookEdit"].includes(toolName)) {
            const filePath = toolInput.file_path || toolInput.notebook_path;
            if (filePath) {
              modifiedFiles.add(filePath);
            }
          }

          // Add tool call as message
          stats.messages.push({
            role: "assistant",
            content: `[Tool: ${toolName}]`,
            timestamp,
            tool_calls: { name: toolName, input: toolInput },
          });
        } else if (msgType === "tool_result") {
          // Add tool result as message (truncated for storage)
          const resultContent = typeof entry.content === "string"
            ? entry.content.slice(0, 2000)
            : JSON.stringify(entry.content || {}).slice(0, 2000);
          stats.messages.push({
            role: "tool",
            content: resultContent,
            timestamp,
            tool_results: { name: entry.name },
          });
        }
      } catch {
        continue;
      }
    }

    // Calculate duration
    if (firstTimestamp && lastTimestamp) {
      stats.duration = Math.round((lastTimestamp.getTime() - firstTimestamp.getTime()) / 1000);
    }

    stats.filesModified = Array.from(modifiedFiles);
  } catch {
    // Ignore parsing errors
  }

  return stats;
}

async function saveFullTranscript(
  sessionId: string,
  stats: TranscriptStats,
  reason: string
): Promise<{ success: boolean; message: string }> {
  if (!API_KEY) {
    return { success: false, message: "No API key configured" };
  }

  if (stats.messages.length === 0) {
    return { success: false, message: "No messages to save" };
  }

  const payload: Record<string, unknown> = {
    session_id: sessionId,
    messages: stats.messages,
    started_at: stats.startedAt,
    source_type: "session_end",
    title: `Session transcript (${reason})`,
    metadata: {
      reason,
      tool_call_count: stats.toolCallCount,
      files_modified: stats.filesModified.slice(0, 20),
      duration_seconds: stats.duration,
    },
    tags: ["session_end", reason],
  };

  if (WORKSPACE_ID) {
    payload.workspace_id = WORKSPACE_ID;
  }
  if (PROJECT_ID) {
    payload.project_id = PROJECT_ID;
  }

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 15000); // 15s for larger payload

    const response = await fetch(`${API_URL}/api/v1/transcripts`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": API_KEY,
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (response.ok) {
      return { success: true, message: `Transcript saved (${stats.messages.length} messages)` };
    }
    return { success: false, message: `API error: ${response.status}` };
  } catch (error) {
    return { success: false, message: String(error) };
  }
}

async function finalizeSession(
  sessionId: string,
  stats: TranscriptStats,
  reason: string
): Promise<void> {
  if (!API_KEY) return;

  // Save full transcript first (if enabled)
  if (SAVE_TRANSCRIPT && stats.messages.length > 0) {
    await saveFullTranscript(sessionId, stats, reason);
  }

  // Then save the summary event
  const payload: Record<string, unknown> = {
    event_type: "manual_note",
    title: `Session Ended: ${reason}`,
    content: JSON.stringify({
      session_id: sessionId,
      reason,
      stats: {
        messages: stats.messageCount,
        tool_calls: stats.toolCallCount,
        duration_seconds: stats.duration,
        files_modified: stats.filesModified.length,
      },
      files_modified: stats.filesModified.slice(0, 20),
      ended_at: new Date().toISOString(),
    }),
    importance: "low",
    tags: ["session", "end", reason],
    source_type: "hook",
  };

  if (WORKSPACE_ID) {
    payload.workspace_id = WORKSPACE_ID;
  }

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 5000);

    await fetch(`${API_URL}/api/v1/memory/events`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": API_KEY,
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);
  } catch {
    // Ignore errors - session is ending anyway
  }
}

export async function runSessionEndHook(): Promise<void> {
  if (!ENABLED) {
    process.exit(0);
  }

  // Read stdin
  let inputData = "";
  for await (const chunk of process.stdin) {
    inputData += chunk;
  }

  if (!inputData.trim()) {
    process.exit(0);
  }

  let input: HookInput;
  try {
    input = JSON.parse(inputData);
  } catch {
    process.exit(0);
  }

  const cwd = input.cwd || process.cwd();
  loadConfigFromMcpJson(cwd);

  const sessionId = input.session_id || "unknown";
  const transcriptPath = input.transcript_path || "";
  const reason = input.reason || "user_exit";

  // Parse transcript for stats
  const stats = parseTranscriptStats(transcriptPath);

  // Finalize session (async, don't wait too long)
  await finalizeSession(sessionId, stats, reason);

  // No output - silent finalization
  process.exit(0);
}

// Auto-run if executed directly
const isDirectRun = process.argv[1]?.includes("session-end") || process.argv[2] === "session-end";
if (isDirectRun) {
  runSessionEndHook().catch(() => process.exit(0));
}

File: crates/mcp-server/src/hook_handlers/session_end.rs
//! SessionEnd (Stop) hook handler.
//!
//! Saves the full transcript when a session ends.
//! This captures the final exchange that the lagging UserPromptSubmit hook misses.

use anyhow::Result;
use serde_json::Value;
use std::collections::HashSet;
use std::path::Path;

use super::{read_stdin_json, write_stdout_json, HookOutput};

/// Handle the Stop/SessionEnd hook.
pub async fn handle() -> Result<()> {
    if std::env::var("CONTEXTSTREAM_SESSION_END_ENABLED")
        .map(|v| v == "false")
        .unwrap_or(false)
    {
        // Silent exit - no output needed for Stop hooks
        write_stdout_json(&HookOutput::empty())?;
        return Ok(());
    }

    let input = read_stdin_json()?;

    let session_id = input
        .get("session_id")
        .and_then(|v| v.as_str())
        .or_else(|| input.get("trajectory_id").and_then(|v| v.as_str()))
        .or_else(|| input.get("execution_id").and_then(|v| v.as_str()))
        .unwrap_or("unknown");
    let transcript_path = input
        .get("transcript_path")
        .and_then(|v| v.as_str())
        .or_else(|| {
            input
                .get("tool_info")
                .and_then(|v| v.get("transcript_path"))
                .and_then(|v| v.as_str())
        })
        .unwrap_or("");
    let reason = input
        .get("reason")
        .and_then(|v| v.as_str())
        .or_else(|| input.get("agent_action_name").and_then(|v| v.as_str()))
        .unwrap_or("user_exit");
    let last_assistant_message = input
        .get("last_assistant_message")
        .and_then(|v| v.as_str())
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(String::from);
    let cwd = input
        .get("cwd")
        .and_then(|v| v.as_str())
        .or_else(|| {
            input
                .get("tool_info")
                .and_then(|v| v.get("cwd"))
                .and_then(|v| v.as_str())
        })
        .map(String::from)
        .or_else(|| {
            std::env::current_dir()
                .ok()
                .and_then(|p| p.to_str().map(String::from))
        })
        .unwrap_or_default();

    let config = load_config(&cwd);

    if config.api_key.is_empty() {
        write_stdout_json(&HookOutput::empty())?;
        return Ok(());
    }

    // Parse transcript
    let mut stats = if !transcript_path.is_empty() && Path::new(transcript_path).exists() {
        parse_transcript_stats(transcript_path)
    } else {
        TranscriptStats::default()
    };

    if stats.messages.is_empty() {
        if let Some(message) = last_assistant_message {
            stats.messages.push(serde_json::json!({
                "role": "assistant",
                "content": message,
                "timestamp": chrono::Utc::now().to_rfc3339(),
            }));
        }
    }

    // Save transcript if enabled (disabled by default).
    // Priority:
    // 1) Hook payload override (`transcripts_enabled` / `save_transcript`)
    // 2) Environment / .mcp.json policy
    let save_transcript = input
        .get("transcripts_enabled")
        .and_then(|v| v.as_bool())
        .or_else(|| input.get("save_transcript").and_then(|v| v.as_bool()))
        .unwrap_or(config.save_transcript_default);

    if save_transcript && !stats.messages.is_empty() {
        let _ = save_full_transcript(&config, session_id, &stats, reason).await;
    }

    // Save summary event
    let _ = save_summary_event(&config, session_id, &stats, reason).await;

    // Silent exit - no output for Stop hooks
    write_stdout_json(&HookOutput::empty())?;
    Ok(())
}

// ============================================================================
// Config Loading
// ============================================================================

struct ApiConfig {
    api_key: String,
    api_url: String,
    workspace_id: Option<String>,
    project_id: Option<String>,
    save_transcript_default: bool,
}

fn parse_bool(value: &str) -> Option<bool> {
    match value.trim().to_ascii_lowercase().as_str() {
        "1" | "true" | "yes" | "on" => Some(true),
        "0" | "false" | "no" | "off" => Some(false),
        _ => None,
    }
}

fn read_mcp_json_env_bool(path: &Path, key: &str) -> Option<bool> {
    let content = std::fs::read_to_string(path).ok()?;
    let config: Value = serde_json::from_str(&content).ok()?;
    let value = config
        .get("mcpServers")?
        .get("contextstream")?
        .get("env")?
        .get(key)?
        .as_str()?;
    parse_bool(value)
}

fn load_config(cwd: &str) -> ApiConfig {
    let mut api_key = std::env::var("CONTEXTSTREAM_API_KEY").unwrap_or_default();
    let mut api_url = std::env::var("CONTEXTSTREAM_API_URL")
        .unwrap_or_else(|_| "https://api.contextstream.io".to_string());
    let mut workspace_id = std::env::var("CONTEXTSTREAM_WORKSPACE_ID").ok();
    let mut project_id: Option<String> = None;
    let mut save_transcript_default = std::env::var("CONTEXTSTREAM_SESSION_END_SAVE_TRANSCRIPT")
        .ok()
        .and_then(|v| parse_bool(&v))
        .or_else(|| {
            std::env::var("CONTEXTSTREAM_HOOK_TRANSCRIPTS_ENABLED")
                .ok()
                .and_then(|v| parse_bool(&v))
        })
        .or_else(|| {
            std::env::var("CONTEXTSTREAM_TRANSCRIPTS_ENABLED")
                .ok()
                .and_then(|v| parse_bool(&v))
        });

    let mut search_dir = std::path::PathBuf::from(cwd);
    for _ in 0..5 {
        let mcp_path = search_dir.join(".mcp.json");
        if api_key.is_empty() {
            if let Some((key, url)) = read_mcp_json_credentials(&mcp_path) {
                api_key = key;
                if let Some(u) = url {
                    api_url = u;
                }
            }
        }
        if save_transcript_default.is_none() {
            save_transcript_default =
                read_mcp_json_env_bool(&mcp_path, "CONTEXTSTREAM_SESSION_END_SAVE_TRANSCRIPT")
                    .or_else(|| {
                        read_mcp_json_env_bool(&mcp_path, "CONTEXTSTREAM_HOOK_TRANSCRIPTS_ENABLED")
                    })
                    .or_else(|| {
                        read_mcp_json_env_bool(&mcp_path, "CONTEXTSTREAM_TRANSCRIPTS_ENABLED")
                    });
        }
        if workspace_id.is_none() || project_id.is_none() {
            let cs_config = search_dir.join(".contextstream").join("config.json");
            if let Ok(content) = std::fs::read_to_string(&cs_config) {
                if let Ok(cfg) = serde_json::from_str::<Value>(&content) {
                    if workspace_id.is_none() {
                        workspace_id = cfg
                            .get("workspace_id")
                            .and_then(|w| w.as_str())
                            .map(String::from);
                    }
                    if project_id.is_none() {
                        project_id = cfg
                            .get("project_id")
                            .and_then(|p| p.as_str())
                            .map(String::from);
                    }
                }
            }
        }
        if !search_dir.pop() {
            break;
        }
    }

    if api_key.is_empty() {
        if let Some(home) = dirs::home_dir() {
            let home_mcp = home.join(".mcp.json");
            if let Some((key, url)) = read_mcp_json_credentials(&home_mcp) {
                api_key = key;
                if let Some(u) = url {
                    api_url = u;
                }
            }
            if save_transcript_default.is_none() {
                save_transcript_default =
                    read_mcp_json_env_bool(&home_mcp, "CONTEXTSTREAM_SESSION_END_SAVE_TRANSCRIPT")
                        .or_else(|| {
                            read_mcp_json_env_bool(
                                &home_mcp,
                                "CONTEXTSTREAM_HOOK_TRANSCRIPTS_ENABLED",
                            )
                        })
                        .or_else(|| {
                            read_mcp_json_env_bool(&home_mcp, "CONTEXTSTREAM_TRANSCRIPTS_ENABLED")
                        });
            }
        }
    }

    ApiConfig {
        api_key,
        api_url,
        workspace_id,
        project_id,
        save_transcript_default: save_transcript_default.unwrap_or(false),
    }
}

fn read_mcp_json_credentials(path: &Path) -> Option<(String, Option<String>)> {
    let content = std::fs::read_to_string(path).ok()?;
    let config: Value = serde_json::from_str(&content).ok()?;
    let env = config.get("mcpServers")?.get("contextstream")?.get("env")?;
    let key = env
        .get("CONTEXTSTREAM_API_KEY")
        .and_then(|k| k.as_str())?
        .to_string();
    let url = env
        .get("CONTEXTSTREAM_API_URL")
        .and_then(|u| u.as_str())
        .map(String::from);
    Some((key, url))
}

// ============================================================================
// Transcript Parsing
// ============================================================================

#[derive(Default)]
struct TranscriptStats {
    #[allow(dead_code)]
    message_count: usize,
    tool_call_count: usize,
    duration_seconds: i64,
    files_modified: Vec<String>,
    messages: Vec<Value>,
    started_at: String,
}

fn parse_transcript_stats(transcript_path: &str) -> TranscriptStats {
    let content = match std::fs::read_to_string(transcript_path) {
        Ok(c) => c,
        Err(_) => return TranscriptStats::default(),
    };

    let mut message_count = 0usize;
    let mut tool_call_count = 0usize;
    let mut modified_files = HashSet::new();
    let mut messages: Vec<Value> = Vec::new();
    let mut first_timestamp: Option<chrono::DateTime<chrono::FixedOffset>> = None;
    let mut last_timestamp: Option<chrono::DateTime<chrono::FixedOffset>> = None;
    let mut started_at = String::new();

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let entry: Value = match serde_json::from_str(line) {
            Ok(v) => v,
            Err(_) => continue,
        };

        let msg_type = entry.get("type").and_then(|t| t.as_str()).unwrap_or("");
        let timestamp = entry
            .get("timestamp")
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_string();

        // Track timestamps
        if let Ok(ts) = chrono::DateTime::parse_from_rfc3339(&timestamp) {
            if first_timestamp.is_none() || ts < first_timestamp.unwrap() {
                first_timestamp = Some(ts);
                started_at = timestamp.clone();
            }
            if last_timestamp.is_none() || ts > last_timestamp.unwrap() {
                last_timestamp = Some(ts);
            }
        }

        match msg_type {
            "user" => {
                message_count += 1;
                if let Some(content) = extract_text(entry.get("content")) {
                    if !content.is_empty() {
                        messages.push(serde_json::json!({
                            "role": "user",
                            "content": content,
                            "timestamp": timestamp,
                        }));
                    }
                }
            }
            "assistant" => {
                message_count += 1;
                if let Some(content) = extract_text(entry.get("content")) {
                    if !content.is_empty() {
                        messages.push(serde_json::json!({
                            "role": "assistant",
                            "content": content,
                            "timestamp": timestamp,
                        }));
                    }
                }
            }
            "tool_use" => {
                tool_call_count += 1;
                let tool_name = entry
                    .get("name")
                    .and_then(|n| n.as_str())
                    .unwrap_or("")
                    .to_string();
                let tool_input = entry.get("input").cloned().unwrap_or(serde_json::json!({}));

                // Track file modifications
                if matches!(tool_name.as_str(), "Write" | "Edit" | "NotebookEdit") {
                    if let Some(fp) = tool_input
                        .get("file_path")
                        .or_else(|| tool_input.get("notebook_path"))
                        .and_then(|p| p.as_str())
                    {
                        modified_files.insert(fp.to_string());
                    }
                }

                messages.push(serde_json::json!({
                    "role": "assistant",
                    "content": format!("[Tool: {}]", tool_name),
                    "timestamp": timestamp,
                    "tool_calls": { "name": tool_name, "input": tool_input },
                }));
            }
            "tool_result" => {
                let result_content = entry
                    .get("content")
                    .map(|c| {
                        if let Some(s) = c.as_str() {
                            s[..s.len().min(2000)].to_string()
                        } else {
                            let s = c.to_string();
                            s[..s.len().min(2000)].to_string()
                        }
                    })
                    .unwrap_or_default();
                messages.push(serde_json::json!({
                    "role": "tool",
                    "content": result_content,
                    "timestamp": timestamp,
                    "tool_results": { "name": entry.get("name").and_then(|n| n.as_str()).unwrap_or("") },
                }));
            }
            _ => {}
        }
    }

    let duration_seconds = match (first_timestamp, last_timestamp) {
        (Some(first), Some(last)) => (last - first).num_seconds(),
        _ => 0,
    };

    if started_at.is_empty() {
        started_at = chrono::Utc::now().to_rfc3339();
    }

    TranscriptStats {
        message_count,
        tool_call_count,
        duration_seconds,
        files_modified: modified_files.into_iter().collect(),
        messages,
        started_at,
    }
}

fn extract_text(content: Option<&Value>) -> Option<String> {
    let value = content?;

    if let Some(text) = value.as_str() {
        return Some(text.to_string());
    }

    if let Some(array) = value.as_array() {
        let mut parts = Vec::new();
        for item in array {
            if let Some(text) = item.get("text").and_then(|entry| entry.as_str()) {
                parts.push(text.to_string());
            } else if let Some(text) = item.as_str() {
                parts.push(text.to_string());
            }
        }
        if !parts.is_empty() {
            return Some(parts.join("\n"));
        }
    }

    if let Some(object) = value.as_object() {
        if let Some(text) = object.get("text").and_then(|entry| entry.as_str()) {
            return Some(text.to_string());
        }
    }

    None
}

// ============================================================================
// API Saves
// ============================================================================

async fn save_full_transcript(
    config: &ApiConfig,
    session_id: &str,
    stats: &TranscriptStats,
    reason: &str,
) -> Result<()> {
    let mut payload = serde_json::json!({
        "session_id": session_id,
        "messages": stats.messages,
        "started_at": stats.started_at,
        "source_type": "session_end",
        "title": format!("Session transcript ({})", reason),
        "metadata": {
            "reason": reason,
            "tool_call_count": stats.tool_call_count,
            "files_modified": &stats.files_modified[..stats.files_modified.len().min(20)],
            "duration_seconds": stats.duration_seconds,
        },
        "tags": ["session_end", reason],
    });

    if let Some(ref ws_id) = config.workspace_id {
        payload
            .as_object_mut()
            .unwrap()
            .insert("workspace_id".to_string(), Value::String(ws_id.clone()));
    }
    if let Some(ref proj_id) = config.project_id {
        payload
            .as_object_mut()
            .unwrap()
            .insert("project_id".to_string(), Value::String(proj_id.clone()));
    }

    let client = super::api_http_client();
    let _ = client
        .post(format!("{}/api/v1/transcripts", config.api_url))
        .header("X-API-Key", &config.api_key)
        .header("Content-Type", "application/json")
        .json(&payload)
        .timeout(std::time::Duration::from_secs(15))
        .send()
        .await;

    Ok(())
}

async fn save_summary_event(
    config: &ApiConfig,
    session_id: &str,
    stats: &TranscriptStats,
    reason: &str,
) -> Result<()> {
    let summary_content = serde_json::json!({
        "session_id": session_id,
        "reason": reason,
        "stats": {
            "messages": stats.messages.len(),
            "tool_calls": stats.tool_call_count,
            "duration_seconds": stats.duration_seconds,
            "files_modified": stats.files_modified.len(),
        },
        "files_modified": &stats.files_modified[..stats.files_modified.len().min(20)],
        "ended_at": chrono::Utc::now().to_rfc3339(),
    });

    let mut payload = serde_json::json!({
        "event_type": "uncategorized",
        "title": format!("Session Ended: {}", reason),
        "content": summary_content.to_string(),
        "importance": "low",
        "tags": ["session", "end", reason],
        "source_type": "hook",
    });

    if let Some(ref ws_id) = config.workspace_id {
        payload
            .as_object_mut()
            .unwrap()
            .insert("workspace_id".to_string(), Value::String(ws_id.clone()));
    }

    let client = super::api_http_client();
    let _ = client
        .post(format!("{}/api/v1/memory/events", config.api_url))
        .header("X-API-Key", &config.api_key)
        .header("Content-Type", "application/json")
        .json(&payload)
        .timeout(std::time::Duration::from_secs(5))
        .send()
        .await;

    Ok(())
}
